File: examples/asyncSendReceive.c
#include "rockblock_9704.h"
#include <string.h>
#include <signal.h>
#include <getopt.h>
#include <limits.h>
#include "crossplatform.h"

#if defined(_WIN32)
#include <io.h>
#define access _access
#else
#include <unistd.h>
#endif

/**
 * This script showcases the asynchronous and queuing capabilites of this library by setting
 * the necessary callbacks, queuing and sending 5 messages and listening for any incoming messages
 * without blocking.
 * 
 * At the start of the script we setup our 4 user defined callbacks to get message provisioning,
 * check if our queued messages have sent, check if we received any messages and finally check the
 * signal strength. A serial connection will then be attempted on the selected port. The script will call
 * rbPoll() everytime it loops, it is important that this is done quite frequently (10ms in this example)
 * as that function is responsible for listening to all the replies from the modem. Five messages will be
 * queued at the start and should start sending if a good signal has been obtained. The callback for message
 * provisioning should then return all the provisioned topics, followed by the onMoComplete callback for 
 * every message sent. The onConstellationState callback will be called anytime either the signal level or
 * signal bars have changed although in this example we will only print the change in signal bars. Finally
 * the script will constantly wait for incoming messages, by waiting for the onMtComplete callback to be called,
 * then store and acknowledge them to clear space for others.
 * 
 * Requirements:
 * RB9704 needs to be provisioned for messaging topic 244 (RAW).
 * Have an open view of the sky where a good signal can be obtained.
 * Adjust MO_QUEUE_SIZE in imt_queue.h to 5U.
 * 
 * (OPTIONAL) If you want to use and test MT queuing do the following:
 * Adjust MT_QUEUE_SIZE in imt_queue.h to desired queue size.
 * Remove or adjust rbAcknowledgeReceiveHeadAsync() so that messages aren't acknowledged right away.
 * Use rbReceiveLockAsync() if you want to lock the queue so that no messages are removed when queue is full.
 * Use rbReceiveUnlockAsync() if you want to unlock the queue.
 * 
*/

static char _serialDevice[PATH_MAX];
static volatile bool _run = true;

int messagesSent = 0;
int currentSignal = 0;
bool receivedNewMessage = false;

typedef enum
{
    SUCCESS = 0,
    INVALID_DEVICE,
    INVALID_ARGUMENTS,
    FAILED_TO_END_CONNECTION,
    FAILED_QUEUE_MESSAGE,
    FAILED_INIT_SERIAL,
} returnCode_t;

static struct option _longOptions[] =
{
    {"device", required_argument, 0, 'd'},
    {"help", no_argument, 0, 'h'},
    {0, 0, 0, 0}
};

static void printHelp(const char * progName)
{
    printf("Usage: %s -d <device> [-h]\n", progName);
    printf("  -d, --device   Serial device to use (mandatory)\n");
    printf("  -h, --help     Display this help message\n");
}

static void signal_handler(int sig)
{
    (void)sig;
    _run = false;
}

void onMessageProvisioning(const jsprMessageProvisioning_t *messageProvisioning)
{
    if(messageProvisioning->provisioningSet == true)
    {
        printf("\033[1;32mDevice is provisioned for %d topics\033[0m\r\n", messageProvisioning->topicCount);
        printf("\033[1;32mProvisioned topics:\033[0m\r\n");
        for(int i = 0; i < messageProvisioning->topicCount; i++)
        {
            printf("\033[1;32mTopic name: %s Topic number: %d\033[0m\r\n", 
            messageProvisioning->provisioning[i].topicName, messageProvisioning->provisioning[i].topicId);
        }
    }
}

void onMoComplete(const unsigned int id, const int status)
{
    printf("\033[1;32mMO Complete: ID = %u, Status = %d\033[0m\r\n", id, status);
    if(status == 1)
    {
        messagesSent += 1;
    }
}

void onMtComplete(const unsigned int id, const int status)
{
    printf("\033[1;32mMT Complete: ID = %u, Status = %d\033[0m\r\n", id, status);
    if(status == 1)
    {
        receivedNewMessage = true;
    }
}

void onConstellationState(const jsprConstellationState_t *state)
{
    if(state->signalBars != currentSignal)
    {
        printf("\033[1;34mCurrent Signal: %d\033[0m\r\n", state->signalBars);
        currentSignal = state->signalBars;
    }
}

int main(int argc, char * argv[])
{
    returnCode_t rVal = SUCCESS;
    bool gotArgs = true;
    int opt = 0;
    _serialDevice[0] = '\0';
    char * mtBuffer = NULL;
    char mtStore[100000];
    int messagesQueued = 0;
    int messagesReceived = 0;
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);

    while ((opt = getopt_long(argc, argv, "d:h", _longOptions, NULL)) != -1)
    {
        switch (opt)
        {
            case 'd':
                strncpy(_serialDevice, optarg, PATH_MAX -1);
            break;

            case 'h':
                printHelp(argv[0]);
                gotArgs = false;
            break;

            case '?':
            // fall through
            default:
                printHelp(argv[0]);
                rVal = INVALID_ARGUMENTS;
            break;
        }
    }
    if ((rVal == SUCCESS) && (gotArgs == true))
    {
        //Assign callbacks
        rbCallbacks_t myCallbacks =
        {
        .messageProvisioning = onMessageProvisioning,
        .moMessageComplete = onMoComplete,
        .mtMessageComplete = onMtComplete,
        .constellationState = onConstellationState
        };
        //Register Callbacks
        rbRegisterCallbacks(&myCallbacks);
        //Begin serial connection and initialise the modem
        if(rbBegin(_serialDevice))
        {
            printf("Successfully started serial session with RB9704\r\n");
            //Queue and send 5 messages
            const char *message = "Message 0";
            const char *message1 = "Message 1";
            const char *message2 = "Message 2";
            const char *message3 = "Message 3";
            const char *message4 = "Message 4";
            //Start listening for MT
            while(_run)
            {
                rbPoll();
                if(messagesQueued == 0)
                {
                    if(rbSendMessageAsync(RAW_TOPIC, message, strlen(message)))
                    {
                        printf("\033[1;32mQueued MO: %s\033[0m\r\n", message);
                        messagesQueued += 1;
                    }
                    
                    if(rbSendMessageAsync(RAW_TOPIC, message1, strlen(message1)))
                    {
                        printf("\033[1;32mQueued MO: %s\033[0m\r\n", message1);
                        messagesQueued += 1;
                    }
                    
                    if(rbSendMessageAsync(RAW_TOPIC, message2, strlen(message2)))
                    {
                        printf("\033[1;32mQueued MO: %s\033[0m\r\n", message2);
                        messagesQueued += 1;
                    }
                    if(rbSendMessageAsync(RAW_TOPIC, message3, strlen(message3)))
                    {
                        printf("\033[1;32mQueued MO: %s\033[0m\r\n", message3);
                        messagesQueued += 1;
                    }
                    if(rbSendMessageAsync(RAW_TOPIC, message4, strlen(message4)))
                    {
                        printf("\033[1;32mQueued MO: %s\033[0m\r\n", message4);
                        messagesQueued += 1;
                    }
                    printf("\033[1;33mMessage Queued: %d\033[0m\r\n", messagesQueued);
                    printf("\033[1;33mMessage Sent: %d\033[0m\r\n", messagesSent);
                }
                //Wait for MT callback then get the message from the head of that queue
                if(receivedNewMessage)
                {
                    receivedNewMessage = false;
                    const size_t mtLength = rbReceiveMessageAsync(&mtBuffer);
                    printf("\033[1;33mMessage length: %ld\033[0m\r\n", mtLength);
                    if ((mtLength > 0) && (mtBuffer != NULL))
                    {
                        //Store the message in another location as we will be clearing it from the queue
                        memcpy(mtStore, mtBuffer, mtLength);
                        printf("Received MT: ");
                        for (size_t i = 0; i < mtLength; i++)
                        {
                            // Print ascii characters
                            if ((mtStore[i] >= 32) && (mtStore[i] <= 126))
                            {
                                printf("\033[1;32m%c\033[0m", mtStore[i]);
                            }
                            else
                            {
                                printf(".");
                            }
                        }
                        printf("\r\n");
                        memset(mtStore, 0, sizeof(mtStore));
                        messagesReceived += 1;
                        printf("\033[1;33mMessages Received: %d\033[0m\r\n", messagesReceived);
                        //Clear the message from the queue to make space
                        if(rbAcknowledgeReceiveHeadAsync())
                        {
                            printf("\033[1;34mMessage acknowledged\033[0m\r\n");
                        }
                    }
                }
                //put cool app stuff here!
                
                //rbPoll() needs to be called very frequently
                usleep(10000);
            }

            //End serial connection
            if(rbEnd())
            {
                printf("Ended connection successfully\r\n");
                _run = false;
            }
            else
            {
                printf("Failed to end connection\r\n");
                rVal = FAILED_TO_END_CONNECTION;
                _run = false;
            }
        }
        else
        {
            printf("Failed to begin the serial connection\r\n");
            rVal = FAILED_INIT_SERIAL;
            _run = false;
        }
    }

    return rVal;
}

File: examples/python/async_send_receive.py
import argparse
from rockblock9704 import *
from time import sleep

#This example code showcases the asynchronous capabilities of this library by setting
#the necessary callbacks, queuing and sending a message, listening for any incoming message
#without blocking and finally quitting after receiving a successful message.
#
#At the start of the script we setup our 4 user defined callbacks to get message provisioning,
#check if our queued message has sent, check if we received any messages and finally check the
#signal strength. A serial connection will then be attempted on the selected port. The script will call
#poll() every time it loops, it is important that this is done quite frequently (10ms is reliable)
#as that function is responsible for listening to all the replies from the modem. A message will be
#queued at the start and should start sending if a good signal has been obtained. The callback for message
#provisioning should then return all the provisioned topics, followed by the onMoComplete callback when 
#the message has sent. The onConstellationState callback will be called anytime either the signal level or
#signal bars have changed although in this example we will only print the change in signal bars. Finally
#the script will constantly wait for an incoming message, by waiting for the onMtComplete callback to be called,
#then store and acknowledge it to clear space.
#
#Requirements:
#RB9704 needs to be provisioned for messaging topic 244 (RAW).
#Have an open view of the sky where a good signal can be obtained.

messages_sent = 0
messages_received = 0
received_new_message = False
current_signal = 0

def on_provision(messageProvisioning):
    if messageProvisioning["provisioningSet"]:
        print(f"\033[1;32mProvisioned for {messageProvisioning['topicCount']} topics\033[0m")
        for topic in messageProvisioning["provisioning"]:
            print(f"\033[1;32mTopic name: {topic['topicName']}, Topic number: {topic['topicId']}\033[0m")

def on_mo(id, status):
    global messages_sent
    print(f"\033[1;32mMO complete: ID={id}, status={status}\033[0m")
    if(status == 1):
        messages_sent += 1
        print(f"\033[1;33mMessage Sent: {messages_sent}\033[0m")

def on_mt(id, status):
    global received_new_message, messages_received
    print(f"\033[1;32mMT complete: ID={id}, status={status}\033[0m")
    if(status == 1):
        messages_received += 1
        print(f"\033[1;32mMessages received: {messages_received}\033[0m")
        received_new_message = True

def on_signal(state):
    global current_signal
    if state["signalBars"] != current_signal:
        print(f"\033[1;34mCurrent Signal: {state['signalBars']}\033[0m")
        current_signal = state["signalBars"]

if __name__ == '__main__':
    parser = argparse.ArgumentParser(prog='receive-message', description="Python example to send IMT message on RockBLOCK 9704.")
    parser.add_argument("--device",
                        help="Serial port of RockBLOCK 9704")

    args = parser.parse_args()

    if (args.device):

        # Create RockBlock9704 instance
        rb = RockBlock9704()
        rb.set_message_provisioning_callback(message_provisioning=on_provision)
        rb.set_mo_message_complete_callback(mo_message_complete=on_mo)
        rb.set_mt_message_complete_callback(mt_message_complete=on_mt)
        rb.set_constellation_state_callback(constellation_state=on_signal)
        # Begin serial communication
        connected = rb.begin(args.device)

        if connected:

            # Get RockBlock information
            print("\033[1;34mBoard temperature: \t\033[0m", rb.get_board_temp())
            print("\033[1;34mSerial number: \t\033[0m", rb.get_serial_number())
            print("\033[1;34m9704 Modem IMEI: \t\033[0m", rb.get_imei())
            print("\033[1;34m9704 Modem ICCID: \t\033[0m", rb.get_iccid())
            messages_queued = 0
            messages_acknowledged = 0
            while True:
                rb.poll()
                if(messages_queued == 0):
                    # Send message with a custom 60s timeout
                    payload = b"Hello, world!"
                    queue = rb.send_message_async(payload, 244)

                    if queue:
                        messages_queued += 1
                        print(f"\033[1;33mMessages queued: {messages_queued}\033[0m")
                    else:
                        # Timed out
                        sleep(0.1)
                        print("\033[1;31mQueueing failed\033[0m")

                # Check for messages
                if received_new_message:
                    message = rb.receive_message_async()

                    if message is not None:
                        print(f"\033[1;32mReceived message: {message}\033[0m")
                        # Acknowledge the head of the queue
                        rb.acknowledge_receive_head_async()
                        messages_acknowledged += 1
                        print(f"\033[1;32mMessages acknowledged: {messages_acknowledged}\033[0m")

                if messages_acknowledged >= 1:
                    break

            if(rb.end()):
                print("Serial connection terminated successfully")

    else:
        print("Please specify a serial port")
        parser.print_help()
